pkg/kine/drivers/generic/batch.go

```diff
@@ -0,0 +1,206 @@
+
+	b.run()
+
+	for job.runId >= b.runId {
+		b.cv.Wait()
+	}
+
+	if job.err != nil {
+		return nil, job.err
+	}
+
+	return job, nil
+}
+
+// run starts a batching job which will run until queue exaustion.
+// run does not block other goroutine from enqueuing new jobs.
+//
+// It must be called while holding the mu lock.
+func (b *Batch) run() {
+	if b.status == batchNotStarted {
+		b.status = batchStarted
+
+		go func() {
+			b.mu.Lock()
+			defer b.mu.Unlock()
+
+			b.status = batchRunning
+			defer func() { b.status = batchNotStarted }()
+
+			for len(b.queue) > 0 {
+				queue := b.queue
+				b.queue = nil
+
+				b.mu.Unlock()
+				b.execQueue(context.TODO(), queue)
+				b.mu.Lock()
+
+				b.runId++
+				b.cv.Broadcast()
+			}
+		}()
+	}
+}
+
+func (b *Batch) execQueue(ctx context.Context, queue []*batchJob) {
+	if len(queue) == 0 {
+		return // This should never happen.
+	}
+	if len(queue) == 1 {
+		// We don't need to address the error here as it will be
+		// handled by the goroutine waiting for this result
+		queue[0].exec(queue[0].ctx, b.db)
+		return
+	}
+
+	transaction := func() error {
+		tx, err := b.db.BeginTx(ctx, nil)
+		if err != nil {
+			return err
+		}
+		defer tx.Rollback()
+
+		for _, q := range queue {
+			// In the case of SQLITE_FULL SQLITE_IOERR SQLITE_BUSY SQLITE_NOMEM
+			// we should explicitly rollback the whole transaction. In all the other
+			// cases, we could keep going with other queries. However, it is a bit
+			// unclear to me what to do next though as:
+			//  - SQLITE_FULL, SQLITE_IOERR mean that we have problems with the disk
+			//    so, even retrying the batch will not work. We might throttle the
+			//    max batch size, hoping in a checkpoint?
+			// - SQLITE_BUSY should never happen if we manage to get `IMMEDIATE`
+			//   transactions in. Otherwise it only affects the first statement.
+			// - SQLITE_NOMEM, again, we could throttle here? Call a gc collection?
+			// Given the points above, the code below always rolls back the whole
+			// batch. It might seem inefficient, but it should almost never happen.
+			if err := q.exec(ctx, tx); err != nil {
+				return err
+			}
+		}
+
+		return tx.Commit()
+	}
+
+	if err := transaction(); err != nil {
+		for _, q := range queue {
+			q.err = err
+		}
+	}
+}
+
+type batchJob struct {
+	ctx   context.Context
+	query string
+	args  []any
+
+	runId        int64
+	lastInsertId int64
+	rowsAffected int64
+	err          error
+}
+
+var _ sql.Result = &batchJob{}
+
+func (bj *batchJob) exec(ctx context.Context, conn BatchConn) error {
+	select {
+	case <-bj.ctx.Done():
+		bj.err = bj.ctx.Err()
+		return bj.err
+	default:
+		// From this point on, the job is not interruptible anymore
+		// as interrupting would mean that we would be forced to
+		// ROLLBACK the whole transaction.
+	}
+
+	var result sql.Result
+	result, bj.err = conn.ExecContext(ctx, bj.query, bj.args...)
+	if bj.err != nil {
+		return bj.err
+	}
+
+	bj.rowsAffected, bj.err = result.RowsAffected()
```

Comment:
Assigning into `bj.rowsAffected` even when `bj.err != nil` doesn't feel 'transactional,' so I think it could be cleaner to do this on another line